### PR TITLE
Fix potential race condition in progress dialog

### PR DESCRIPTION
**Pull request by Megamouse**

### CODE REVIEW

File: rpcs3/Emu/system_progress.hpp [hunk 1]
```diff
@@ -59,11 +60,12 @@ class scoped_progress_dialog final
 	scoped_progress_dialog& operator=(const std::string& text) noexcept
 	{
 		ensure(!text.empty());
-		m_text = text;
 
 		// This method is destroying the previous value and replacing it with a new one
-		std::tie(m_prev, m_prev_id, m_id) = g_progr_text.atomic_op([this](progress_dialog_string_t& progr)
+		std::tie(m_prev, m_prev_id, m_id) = g_progr_text.atomic_op([this, text](progress_dialog_string_t& progr)
 		{
+			m_text = text;
```
Comment (elad335) [inline rpcs3/Emu/system_progress.hpp:67]: 1) You imply that operator= is called concurrently on the same object? 
2) If 1 is true, then how it fixes it?

File: rpcs3/Emu/system_progress.hpp [hunk 2]
```diff
@@ -38,10 +38,11 @@ class scoped_progress_dialog final
 	scoped_progress_dialog(const std::string& text) noexcept
 	{
 		ensure(!text.empty());
-		m_text = text;
 
-		std::tie(m_prev, m_prev_id, m_id) = g_progr_text.atomic_op([this](progress_dialog_string_t& progr)
+		std::tie(m_prev, m_prev_id, m_id) = g_progr_text.atomic_op([this, text](progress_dialog_string_t& progr)
 		{
+			m_text = text;
```
Comment (elad335) [inline rpcs3/Emu/system_progress.hpp:44]: There is no race in access of data inside a constructor, a constructor (should) occur only before the object the is accessable to everyone else.

**Review by elad335 — CHANGES_REQUESTED** [review]
I think m_text does not have a use anymore, it was meant to compare with the g_progr_text on destructor to see if the value matches before string restoration. But now there is ID. So m_text can be removed.

**Comment by Megamouse** [comment]
m_text is new lol. It's used as storage for the const char* in the atomic object

File: rpcs3/Emu/system_progress.hpp (hunk 2 above)
Comment (Megamouse) [reply rpcs3/Emu/system_progress.hpp:44]: true

**Comment by Megamouse** [comment]
Since we now use translatable strings instead of constexpr const char*

File: rpcs3/Emu/system_progress.hpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/system_progress.hpp:67]: I think I saw instances where it's updated inside worker threads

**Comment by elad335** [comment]
> Since we now use translatable strings instead of constexpr const char*

This is a bug, m_text is an unrelated copy, text.c_str() won't equal to m_text.c_str()
std::string copy is a not shared_ptr copy-alike or something, it doesn't let extend the string's life

File: rpcs3/Emu/system_progress.hpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/system_progress.hpp:67]: Where? this should be a local object, it's not meant for this

**Comment by Megamouse** [comment]
Oops. For some reason I thought the constructor resets the ptr when the thing goes out of scope. It's not hard to fix but I can only do it later

**Comment by elad335** [comment]
It's fixed in #16321, that was the point of the pr

**Pull request closed without merging**